"hass variable not set on template" in 2024.8.b04

### The problem

In 2024.8.0b4 I receive the following error message on all rest integrations calls.

```
2024-08-06 15:16:19.314 ERROR (MainThread) [homeassistant.components.sensor] Error adding entity sensor.group_therapy_episode_2 for domain sensor with platform rest
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 598, in _async_add_entities
await coro
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 912, in _async_add_entity
await entity.add_to_platform_finish()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1365, in add_to_platform_finish
await self.async_added_to_hass()
File "/usr/src/homeassistant/homeassistant/helpers/trigger_template_entity.py", line 162, in async_added_to_hass
await super().async_added_to_hass()
File "/usr/src/homeassistant/homeassistant/components/rest/entity.py", line 43, in async_added_to_hass
self._update_from_rest_data()
File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 187, in _update_from_rest_data
self._process_manual_data(raw_value)
File "/usr/src/homeassistant/homeassistant/helpers/trigger_template_entity.py", line 250, in _process_manual_data
self._render_templates(variables)
File "/usr/src/homeassistant/homeassistant/helpers/trigger_template_entity.py", line 190, in _render_templates
rendered[key] = self._config[key].async_render(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 597, in async_render
compiled = self._compiled or self._ensure_compiled(limited, strict, log_fn)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 798, in _ensure_compiled
assert self.hass is not None, "hass variable not set on template"
^^^^^^^^^^^^^^^^^^^^^
AssertionError: hass variable not set on template
```

I have a packages folder set up as follows in configuration.yaml
```yaml
homeassistant:
  customize: !include customize.yaml
  packages: !include_dir_named packages
```

I also have rest configured as follows with some other data being pulled, if this is relevant.
```yaml
rest: !include rest.yaml
```

Bear in mind this works just fine in 2024.7.4.

In my `packages` folder, I have a file called `podcasts.yaml` which defines a load of entries to retrieve podcasts and data from youtube and also defines some templates. I can't post a copy, it's 1000+ lines but here's an example:

It starts like this:
```yaml
rest:
  - resource: "https://www.youtube.com/feeds/videos.xml?playlist_id=PLhHqebDQf-abSrU9LHEsaLU4P2R6W_RiK"
    scan_interval: 3600
    sensor:
      - name: "lane_8_episode_0"
        unique_id: lane_8_episode_0
        value_template: "{{ value_json.feed.entry[0].title }}"
        picture: "{{ value_json.feed.entry[0]['media:group']['media:thumbnail']['@url'] }}"
        json_attributes_path: "$.feed.entry[0]"
        json_attributes:
          - link
          - published
      - name: "lane_8_episode_1"
        unique_id: lane_8_episode_1
        value_template: "{{ value_json.feed.entry[1].title }}"
        picture: "{{ value_json.feed.entry[1]['media:group']['media:thumbnail']['@url'] }}"
        json_attributes_path: "$.feed.entry[1]"
        json_attributes:
          - link
          - published
      - name: "lane_8_episode_2"
        unique_id: lane_8_episode_2
        value_template: "{{ value_json.feed.entry[2].title }}"
        picture: "{{ value_json.feed.entry[2]['media:group']['media:thumbnail']['@url'] }}"
        json_attributes_path: "$.feed.entry[2]"
        json_attributes:
          - link
          - published
      - name: "lane_8_episode_3"
        unique_id: lane_8_episode_3
        value_template: "{{ value_json.feed.entry[3].title }}"
        picture: "{{ value_json.feed.entry[3]['media:group']['media:thumbnail']['@url'] }}"
        json_attributes_path: "$.feed.entry[3]"
        json_attributes:
          - link
          - published
      - name: "lane_8_episode_4"
        unique_id: lane_8_episode_4
        value_template: "{{ value_json.feed.entry[4].title }}"
        picture: "{{ value_json.feed.entry[4]['media:group']['media:thumbnail']['@url'] }}"
        json_attributes_path: "$.feed.entry[4]"
        json_attributes:
          - link
          - published
```

Using a template, defined within the same podcasts.yaml file, I use a number helper (which is the thing that changes) to pull data from the rest sensors defined above and populate the template sensor.

```yaml
template:
  - sensor:
      - name: Lane 8
        unique_id: lane_8
        icon: mdi:podcast
        state: "{{ states('sensor.lane_8_episode_'~(states('input_number.podcast_episode') | int(0))) }}"
        attributes:
          link: "{{ state_attr('sensor.lane_8_episode_'~(states('input_number.podcast_episode') | int(0)), 'link')['@href'] }}"
          image: "{{ state_attr('sensor.lane_8_episode_'~(states('input_number.podcast_episode') | int(0)), 'entity_picture') }}"
          published: |
            {% set date = state_attr('sensor.lane_8_episode_'~(states('input_number.podcast_episode') | int(0)), 'published') %}
            {% from 'date_ordinals.jinja' import date_ordinal %}
            {{date_ordinal(as_datetime(date, now()).isoformat(), false, false, true)}}
```

I noticed initially that the image attribute was null in the template sensor but on investigation, it was because the `picture` attribute wasn't on the episode entry anymore. Again, this works just fine in 2024.7.4

My logs have the above error code for every podcast `episode` entry in the podcasts.yaml file but I don't get this in 2024.7.4 or any previous version.

The rest integration entries _are_ created but they do not have the picture attribute (entity_picture) for some reason.
2024.8.b04
![image](https://github.com/user-attachments/assets/14729331-a6fa-4e70-9acf-53b8b968df48)

2024.7.4
![image](https://github.com/user-attachments/assets/40786240-2a5c-421d-8a68-179538e4aa08)


### What version of Home Assistant Core has the issue?

core-2024.8.0b4

### What was the last working version of Home Assistant Core?

core-2024.7.4

### What type of installation are you running?

Home Assistant OS

### Integration causing the issue

REST

### Link to integration documentation on our website

_No response_

### Diagnostics information

_No response_

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

--- a/homeassistant/config.py
+++ b/homeassistant/config.py
@@ -817,9 +817,7 @@ async def async_process_ha_core_config(hass: HomeAssistant, config: dict) -> Non
 
     This method is a coroutine.
     """
-    # CORE_CONFIG_SCHEMA is not async safe since it uses vol.IsDir
-    # so we need to run it in an executor job.
-    config = await hass.async_add_executor_job(CORE_CONFIG_SCHEMA, config)
+    config = CORE_CONFIG_SCHEMA(config)
 
     # Only load auth during startup.
     if not hasattr(hass, "auth"):
@@ -1535,15 +1533,9 @@ async def async_process_component_config(
             return IntegrationConfigInfo(None, config_exceptions)
 
     # No custom config validator, proceed with schema validation
-    if config_schema := getattr(component, "CONFIG_SCHEMA", None):
+    if hasattr(component, "CONFIG_SCHEMA"):
         try:
-            if domain in config:
-                # cv.isdir, cv.isfile, cv.isdevice are not async
-                # friendly so we need to run this in executor
-                schema = await hass.async_add_executor_job(config_schema, config)
-            else:
-                schema = config_schema(config)
-            return IntegrationConfigInfo(schema, [])
+            return IntegrationConfigInfo(component.CONFIG_SCHEMA(config), [])
         except vol.Invalid as exc:
             exc_info = ConfigExceptionInfo(
                 exc,

